shortfin/tests/apps/sd/e2e_test.py

```diff
@@ -0,0 +1,201 @@
+    time.sleep(5)
+
+    yield runner
+
+    # Teardown: kill the server
+    del runner
+
+
+@pytest.mark.system("amdgpu")
+def test_sd_server(sd_server):
+    imgs, status_code = send_json_file(sd_server.url)
+    assert len(imgs) == 1
+    assert status_code == 200
+
+
+class ServerRunner:
+    def __init__(self, args):
+        port = str(find_free_port())
+        self.url = "http://0.0.0.0:" + port
+        env = os.environ.copy()
+        env["PYTHONUNBUFFERED"] = "1"
+        env["HIP_VISIBLE_DEVICES"] = "0"
+        self.process = subprocess.Popen(
+            [
+                *args,
+                "--port=" + port,
+                "--device=amdgpu",
+            ],
+            env=env,
+            # TODO: Have a more robust way of forking a subprocess.
+            stdout=sys.stdout,
+            stderr=sys.stderr,
+        )
+        print(self.process.args)
+        self._wait_for_ready()
+
+    def _wait_for_ready(self):
+        start = time.time()
+        while True:
+            time.sleep(2)
+            try:
+                if requests.get(f"{self.url}/health").status_code == 200:
+                    return
+            except Exception as e:
+                if self.process.errors is not None:
+                    raise RuntimeError("API server processs terminated") from e
+            time.sleep(1.0)
+            if (time.time() - start) > 30:
+                raise RuntimeError("Timeout waiting for server start")
+
+    def __del__(self):
+        try:
+            process = self.process
+        except AttributeError:
+            pass
+        else:
+            process.terminate()
+            process.wait()
+
+
+def bytes_to_img(bytes, idx=0, width=1024, height=1024):
+    timestamp = dt.now().strftime("%Y-%m-%d_%H-%M-%S")
+    image = Image.frombytes(
+        mode="RGB", size=(width, height), data=base64.b64decode(bytes)
+    )
+    return image
+
+
+def send_json_file(url="http://0.0.0.0:8000"):
+    # Read the JSON file
+    data = sample_request
+    imgs = []
+    # Send the data to the /generate endpoint
+    try:
+        response = requests.post(url + "/generate", json=data)
+        response.raise_for_status()  # Raise an error for bad responses
+        request = json.loads(response.request.body.decode("utf-8"))
+
+        for idx, item in enumerate(response.json()["images"]):
+            width = (
+                request["width"][idx]
+                if isinstance(request["height"], list)
+                else request["height"]
+            )
+            height = (
+                request["height"][idx]
+                if isinstance(request["height"], list)
+                else request["height"]
+            )
+            img = bytes_to_img(item.encode("utf-8"), idx, width, height)
+            imgs.append(img)
+
+    except requests.exceptions.RequestException as e:
+        print(f"Error sending the request: {e}")
+
+    return imgs, response.status_code
+
+
+def find_free_port():
```

Comment:
This and the above has been copied a couple of times now. Probably need a test_utils of some kind.